pkg/controller/windows/windows_controller.go

```diff
@@ -0,0 +1,290 @@
+
+	"github.com/tigera/operator/pkg/components"
+	"golang.org/x/mod/semver"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	log = logf.Log.WithName("controller_windows")
+
+	// The Enterprise version this operator supports.
+	supportedUpgradeVersion = fmt.Sprintf("Enterprise-%v", components.EnterpriseRelease)
+)
+
+// Add creates a new Windows Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, opts options.AddOptions) error {
+	if !opts.EnterpriseCRDExists {
+		// No need to start this controller.
+		return nil
+	}
+	// create the reconciler
+	reconciler := newReconciler(mgr, opts)
+
+	// Create a new controller
+	controller, err := controller.New("windows-controller", mgr, controller.Options{Reconciler: reconcile.Reconciler(reconciler)})
+	if err != nil {
+		return err
+	}
+
+	return add(mgr, controller)
+}
+
+// newReconciler returns a new *reconcile.Reconciler
+func newReconciler(mgr manager.Manager, opts options.AddOptions) reconcile.Reconciler {
+	r := &reconcileWindows{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		provider:      opts.DetectedProvider,
+		status:        status.New(mgr.GetClient(), "windows", opts.KubernetesVersion),
+		clusterDomain: opts.ClusterDomain,
+	}
+	r.status.Run()
+	return r
+}
+
+// add adds watches for resources that are available at startup
+func add(mgr manager.Manager, c controller.Controller) error {
+	return c.Watch(&source.Kind{Type: &operatorv1.Installation{}}, &handler.EnqueueRequestForObject{})
+}
+
+// blank assignment to verify that reconcileWindows implements reconcile.Reconciler
+var _ reconcile.Reconciler = &reconcileWindows{}
+
+// reconcileWindows reconciles windows nodes. For now, this handles Calico
+// Windows upgrades.
+type reconcileWindows struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client        client.Client
+	scheme        *runtime.Scheme
+	provider      operatorv1.Provider
+	status        status.StatusManager
+	clusterDomain string
+}
+
+func (r *reconcileWindows) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Windows")
+
+	// Get supported Windows nodes. Nodes are supported if they have a version
+	// annotation.
+	nodes, err := common.GetWindowsNodes(ctx, r.client, func(n *corev1.Node) bool {
```

Comment:
Yep agreed. I don't think we need a dudup mechanism yet since we're filtering the node events we're going to watch, so that only the node events we care about will trigger a reconcile.